lib/src/modules/gates.dart

```diff
@@ -513,3 +513,64 @@ class Mux extends Module with InlineSystemVerilog {
     return '$control ? $d1 : $d0';
   }
 }
+
+/// A generic two-input bit index gate [Module].
+///
+/// It always takes two inputs and has one output of width 1.
+class IndexGate extends Module with InlineSystemVerilog {
+  late final String _a, _b, _y;
+
+  /// The primary input to this gate.
+  Logic get a => input(_a);
+
+  /// The bit index for this gate.
+  Logic get b => input(_b);
+
+  /// The output of this gate.
+  Logic get y => output(_y);
+
+  /// Constructs a two-input bit index gate for an abitrary custom functional implementation.
+  ///
+  /// The bit [_b] will be indexed as an output.
+  /// [Module] is in-lined as SystemVerilog, it will use _a[index], where index is a Const(_b)
+  IndexGate(Logic a, Logic b): super() {
+    _a = 'a_${a.name}';
+    _b = 'b_${b.name}';
```

Comment:
the index/b should be unpreferred since that can be fed in with an in-lined expression, but keep original/a as non-unpreferred.